Diz.Core/model/snes/Data.cs

```diff
@@ -0,0 +1,471 @@
+
+        // NOTE: specially named parameterized constructor for XML serializer
+        [UsedImplicitly] 
+        public Data(ByteSource snesAddressSpace)
+        {
+            SnesAddressSpace = snesAddressSpace;
+        }
+        
+        private byte[] GetRomBytes(int snesOffset, int count)
+        {
+            var output = new byte[count];
+            for (var i = 0; i < output.Length; i++)
+            {
+                var snesAddress = snesOffset + i;
+                output[i] = GetSnesByte(snesAddress) 
+                            ?? throw new InvalidDataException($"No data at SNES offset {snesAddress:X}");
+            }
+
+            return output;
+        }
+
+        // TODO: offset isn't snes it's rom? how is this still able to work? figure it out and fix variable naming
+        public string GetRomNameFromRomBytes() => GetFixedLengthStr(0xFFC0, 21);
+
+        private string GetFixedLengthStr(int snesOffset, int count)
+        {
+            return Encoding.UTF8.GetString(GetRomBytes(snesOffset, count));
+        }
+
+        // TODO: offset isn't snes it's rom? how is this still able to work? figure it out and fix variable naming
+        // TODO: replace with GetRomDoubleWord()
+        public int GetRomCheckSumsFromRomBytes() => ByteUtil.ByteArrayToInt32(GetRomBytes(0xFFDC, 4));
+
+        public int GetRomSize() => RomByteSource?.Bytes?.Count ?? 0;
+        
+        // -------------
+        // probably can move some of this annotation stuff into ByteSource
+        // -------------
+        
+        public T GetOneAnnotationAtPc<T>(int pcOffset) where T : Annotation, new()
+        {
+            var snesAddress = ConvertPCtoSnes(pcOffset);
+            return SnesAddressSpace.GetOneAnnotation<T>(snesAddress);
+        }
+
+        public T GetOrCreateAnnotationAtPc<T>(int pcOffset) where T : Annotation, new()
+        {
+            var snesOffset = ConvertPCtoSnes(pcOffset);
+            return _GetOrCreateAnnotation<T>(snesOffset).annotation;
+        }
+        
+        private (T annotation, bool wasExisting) _GetOrCreateAnnotation<T>(int snesOffset) where T : Annotation, new()
+        {
+            var existing = SnesAddressSpace.GetOneAnnotation<T>(snesOffset);
+            if (existing != null)
+                return (existing, true);
+
+            var newAnnotation = new T();
+
+            // NOTE: for now, we add all annotations on the Snes Address space itself.
+            // in the future, we might want to push them further down into nested ByteSource's like the 
+            // ROM itself, or, the WRAM.  Especially useful because pushing something into a mirror WRAM region
+            // will make the mirroring automagically work out nicely.
+            SnesAddressSpace.AddAnnotation(snesOffset, newAnnotation);
+            
+            return (newAnnotation, false);
+        }
+
+        public FlagType GetFlag(int i) => GetOneAnnotationAtPc<MarkAnnotation>(i)?.TypeFlag ?? default;
```

Comment:
all these helpers are probably areas we could simplify or make the callers do this work